src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -25,37 +25,40 @@ private BoundExpression BindDeconstruction(AssignmentExpressionSyntax node, Diag
             var left = node.Left;
             var right = node.Right;
 
-            if (node.IsDeconstructionDeclaration())
-            {
-                return BindDeconstructionDeclaration(node, left, right, diagnostics);
-            }
-
-            AssertDeconstructionIsAssignment(left);
+            bool isDeclaration = node.IsDeconstructionDeclaration();
+            Debug.Assert(isDeclaration || DeconstructionIsAssignment(left));
+            return BindDeconstruction(node, left, right, diagnostics, isDeclaration);
+        }
 
-            var tuple = (TupleExpressionSyntax)left;
-            ArrayBuilder<DeconstructionVariable> checkedVariables = BindDeconstructionAssignmentVariables(tuple.Arguments, tuple, diagnostics);
-            var result = BindDeconstructionAssignment(node, node.Right, checkedVariables, diagnostics, isDeclaration: false);
-            FreeDeconstructionVariables(checkedVariables);
+        internal BoundDeconstructionAssignmentOperator BindDeconstruction(
+                                                            CSharpSyntaxNode node,
+                                                            ExpressionSyntax left,
+                                                            ExpressionSyntax right,
+                                                            DiagnosticBag diagnostics,
+                                                            bool isDeclaration,
+                                                            BoundDeconstructValuePlaceholder rightPlaceholder = null)
+        {
+            DeconstructionVariable locals = BindDeconstructionVariables(left, isDeclaration, diagnostics);
+            Debug.Assert(locals.HasNestedVariables);
+            var result = BindDeconstructionAssignment(node, right, locals.NestedVariables, diagnostics, rhsPlaceholder: rightPlaceholder);
+            FreeDeconstructionVariables(locals.NestedVariables);
             return result;
         }
 
-        [Conditional("DEBUG")]
-        private void AssertDeconstructionIsAssignment(ExpressionSyntax expression)
+        /// <summary>
+        /// Checks that the expression doesn't contain any declaration expressions.
+        /// </summary>
+        private bool DeconstructionIsAssignment(ExpressionSyntax expression)
         {
             switch (expression.Kind())
             {
                 case SyntaxKind.DeclarationExpression:
-                    Debug.Assert(false);
-                    break;
+                    return false;
                 case SyntaxKind.TupleExpression:
                     var tuple = (TupleExpressionSyntax)expression;
-                    foreach (var arg in tuple.Arguments)
-                    {
-                        AssertDeconstructionIsAssignment(arg.Expression);
-                    }
-                    break;
+                    return tuple.Arguments.All(a => DeconstructionIsAssignment(a.Expression));
```

Comment:
Is this calling `Enumerable.All(IEnumerable<T>, ...)`. If so, consider adding a `SeparatedSyntaxList<T>.All()` method or using `SeparatedSyntaxList<T>.IndexOf()` instead.